File: server/game_core/actor.py
import math

from game_core.game_object import GameObject
from game_core.physics.point import Point
from game_core.physics.rectangle import Rectangle


class Actor(GameObject):
    game = None

    x = None
    y = None
    width = None
    height = None
    rotation = None
    visible = None

    collisionBuffer = None

    def __init__(self, actor_id, width=0, height=0):
        super(Actor, self).__init__(actor_id)

        self.visible = True
        self.collisionBuffer = 10

        self.x = 0
        self.y = 0
        self.width = width
        self.height = height

        self.rotation = 0

    def update(self, delta):
        while self.commands:
            cmd = self.commands.pop(0)

            cmd.execute(self)

        self.act()

    def act(self):
        pass

    def set_position(self, x=0, y=0):
        self.x = x
        self.y = y

    def set_rotation(self, angle):
        self.rotation = angle % 360

    def move(self, speed=1):
        angle = self.rotation * (math.pi / 180.0)
        x = math.trunc((self.x + math.cos(angle) * speed))
        y = math.trunc((self.y + math.sin(angle) * speed))

        self.set_position(x, y)

    def get_bounding_box(self):
        return Rectangle(Point(self.x - self.collisionBuffer, self.y - self.collisionBuffer),
                         Point(self.x + self.width + self.collisionBuffer, self.y + self.height + self.collisionBuffer))

    def intersecting(self, actor):
        return self.get_bounding_box().overlaps(actor.get_bounding_box())

    def intersecting_position(self, x, y):
        return self.get_bounding_box().contains(Point(x, y))

    def at_game_top_edge(self):
        return self.y < self.collisionBuffer

    def at_game_bottom_edge(self):
        return self.y + self.height > self.game.height - self.collisionBuffer

    def at_game_left_edge(self):
        return self.x < self.collisionBuffer

    def at_game_right_edge(self):
        return self.x + self.width > self.game.width - self.collisionBuffer

File: src/client/lib/game/core/abstract/actor.dart
import 'dart:math';
import 'package:client/game/core/assets/game_image.dart';
import 'package:client/game/core/game.dart';
import 'package:client/game/core/abstract/game_object.dart';
import 'package:client/game/core/command.dart';

abstract class Actor extends GameObject
{
  Game game;

  Point position;
  double rotation;
  bool visible;

  GameImage image;
  int collisionBuffer;

  int get X => position.x;
  int get Y => position.y;

  Actor(String ID, [this.position = const Point(0, 0), this.rotation = 0.0]) : super(ID)
  {
    visible = true;
    collisionBuffer = 10;
  }

  @override
  update(num delta)
  {
    while(commands.isNotEmpty)
    {
      Command command = commands.removeFirst();
      command.execute(this);
    }

    act();
  }

  act();

  setPosition(int x, int y)
  {
    position = new Point(x, y);
  }

  setRotation(double angle)
  {
    //Maintain an angle within 360 degrees
    rotation = angle % 360;
  }

  setImageByUrl(String url)
  {
    this.image = new GameImage.fromUrl(url);
  }

  setImage(GameImage image)
  {
    this.image = new GameImage.from(image);
  }

  move(int speed)
  {
    double angle = rotation * (PI / 180.0);
    int x = (X + cos(angle) * speed).round();
    int y = (Y + sin(angle) * speed).round();

    setPosition(x, y);
  }

  Rectangle getBoundingBox()
  {
    //Add a buffer to give some space for collisions
    return new Rectangle(
        X - collisionBuffer,
        Y - collisionBuffer,
        image.width + collisionBuffer,
        image.height + collisionBuffer);
  }

  bool intersecting(Actor other)
  {
    return getBoundingBox().intersects(other.getBoundingBox());
  }

  bool intersectingPosition(Point pos)
  {
    return getBoundingBox().containsPoint(pos);
  }

  atGameTopEdge()
  {
    return Y < collisionBuffer;
  }

  atGameBottomEdge()
  {
    return Y + image.height > game.height - collisionBuffer;
  }

  atGameLeftEdge()
  {
    return X < collisionBuffer;
  }

  atGameRightEdge()
  {
    return X + image.width > game.width - collisionBuffer;
  }
}